src/rosenfeld_fmt.cc

```diff
@@ -669,36 +695,63 @@ void RosenfeldFMT::computeSphericalWeightedDensities(std::shared_ptr<State> stat
-            const auto dr = mesh->step();
-
+            const auto rho = state->getField(t)->const_view(); 
             auto n2i = tmp_field_["n2"]->view();
             auto n3i = tmp_field_["n3"]->view();
             auto nv2i = tmp_field_["nv2"]->view();
-
+            
+            
             for (int idx=0; idx < std::min(mesh->bin(R), mesh->shape()); ++idx)
                 {
                 // reset n for this bin
                 n2i(idx) = 0.;
                 n3i(idx) = 0.;
                 nv2i(idx) = 0.;
-
-                // take integrals using quadrature
+                    
+                // take integrals using trapezoidal rule
                 const auto r = mesh->center(idx);
-                const auto lower_bin = mesh->bin(0.);
-                const auto split_bin = mesh->bin(R-r);
-                const auto upper_bin = mesh->bin(r+R);
-                for (int ig_idx=lower_bin; ig_idx < split_bin; ++ig_idx)
+    
+                const double lower = 0;
+                const double split = R-r;
+                const double upper = r+R;
+                double x ;
+                double n = 100;
+                double dr;
+            
+                //Initial step for the first integral 
+                double n3_ig_1 = 0;
+                double n3_ig_2 = 0; 
+                double n2_ig_2 = 0;
+                double nv2_ig_2 = 0;
+            
+                //Integral from 0 to R-r
+                x = lower;
+                dr = split/n;
+                for (int ig_idx=0; ig_idx <=n; ig_idx++)
                     {
-                    n3i(idx) += 4. * M_PI * r_rho(ig_idx) * mesh->center(ig_idx) * dr;
+                    const double factor = (ig_idx == 0 || ig_idx == n) ? 0.5 : 1.0;
+                    double rho_ig = mesh->interpolate(x, rho);
+                    n3_ig_1 += factor * 4*M_PI* x * x * rho_ig;
+                    x += dr;
                     }
-                for (int ig_idx=split_bin; ig_idx < upper_bin; ++ig_idx)
+                
+                 //Integral from R-r to r+R
+                 x = split;
+                 dr = (upper-split)/n;
+
+                for (int ig_idx=0; ig_idx <=n; ig_idx++)
                     {
-                    const auto rig = mesh->center(ig_idx);
-                    n3i(idx) += (M_PI/r) * r_rho(ig_idx) * (R*R - (r-rig)*(r-rig)) * dr;
-                    n2i(idx) += (2.*M_PI*R/r) * r_rho(ig_idx) * dr;
-                    nv2i(idx) += (M_PI/(r*r)) * r_rho(ig_idx) * (R*R + r*r - rig*rig) * dr;
+                    const double factor = (ig_idx == 0 || ig_idx == n) ? 0.5 : 1.0;
+                    const auto rho_ig = mesh->interpolate(x,rho);
+                    n3_ig_2 += factor * (M_PI/r) * (rho_ig *x* (R*R - (r-x)*(r-x)));
+                    n2_ig_2 += factor * (2.*M_PI*R/r) * rho_ig * x;
+                    nv2_ig_2 += factor * (M_PI/(r*r)) * (rho_ig * x  * (R*R + r*r - x*x));
```

Comment:
Can't these be acculumated directly?
```suggestion
                    n2(idx) += factor * (2.*M_PI*R/r) * rho_ig * x;
                    nv2(idx) += factor * (M_PI/(r*r)) * (rho_ig * x  * (R*R + r*r - x*x));
```